### PR TITLE
TypeSum improvements

### PR DESCRIPTION
**Pull request by alexeyr**

### CODE REVIEW

**Comment by alexeyr** [comment]
@4e6 @alexanderfilippov One more.

File: core/src/main/scala/scalan/primitives/TypeSum.scala [hunk 1]
```diff
@@ -1,216 +1,248 @@
...
+    def unapply[A, B](lam: Lambda[A, B]): Option[SumMap[_, _, _, _]] = lam.y match {
+      case Def(m: SumMap[_, _, _, _]) if lam.x == m.sum => Some(m)
       case _ => None
     }
   }
 
-  def liftFromSumFold[T1,T2,A,B](
-        sum: Rep[T1|T2], left: Rep[T1 => B], right: Rep[T2 => B], iso: Iso[A,B]): Rep[B] = {
+  def liftFromSumFold[T1, T2, A, B](
+                                     sum: Rep[T1 | T2], left: Rep[T1 => B], right: Rep[T2 => B], iso: Iso[A, B]): Rep[B] = {
```
Comment (4e6) [inline core/src/main/scala/scalan/primitives/TypeSum.scala]: Formatting. I'd rather join with previous line.

File: core/src/main/scala/scalan/primitives/TypeSum.scala [hunk 2]
```diff
@@ -1,216 +1,248 @@
...
     val res = sum.foldBy(iso.fromFun << left, iso.fromFun << right)
     iso.to(res)
   }
 
-  def liftFromSumFold[T1,T2,A,B,C,D](
-        sum: Rep[T1|T2], left: Rep[T1 => C], right: Rep[T2 => D],
-        iso1: Iso[A,C], iso2: Iso[B,D],
-        toD: Conv[C,D], toC: Conv[D,C]): Rep[C] = {
+  def liftFromSumFold[T1, T2, A, B, C, D](
+                                           sum: Rep[T1 | T2], left: Rep[T1 => C], right: Rep[T2 => D],
```
Comment (4e6) [inline core/src/main/scala/scalan/primitives/TypeSum.scala]: The same as above.

**Comment by 4e6** [comment]
I anticipate a large refactoring of my `Iterators` package :) 
:+1:

**Comment by alexeyr** [comment]
@4e6 You may want to follow #64 then. Or even do it yourself :)

**Pull request merged**